Data race in CI visibility and orchestrion

> Hey @tonyredondo, I also recently tried introducing CI visibility and `orchestrion` to our Go pipelines and hit some similar looking errors. From a glance, our race failures were also in table tests.
> 
> Do you think these would be addressed by your fix?
> 
> <details>
>   <summary>Sample error log</summary>
>   
>   ```
>   --- FAIL: TestGetAttrs (0.01s)
>       --- FAIL: TestGetAttrs/nil_error (0.00s)
>           testing.go:1490: race detected during execution of test
>   FAIL	github.com/<redacted>	0.843s
>   ==================
>   WARNING: DATA RACE
>   Read at 0x00000315a608 by main goroutine:
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations/gotesting.applyAdditionalFeaturesToTestFunc()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/gotesting/instrumentation.go:159 +0x58
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations/gotesting.(*M).executeInternalTest()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/gotesting/testing.go:354 +0x32f
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations/gotesting.(*M).instrumentInternalTests()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/gotesting/testing.go:158 +0xb64
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations/gotesting.instrumentTestingM()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/gotesting/instrumentation_orchestrion.go:58 +0x204
>     testing.(*M).Run()
>         <generated>:2 +0x4a
>     main.main()
>         _testmain.go:45 +0x164
>   
>   Previous write at 0x00000315a608 by goroutine 75:
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations.ensureAdditionalFeaturesInitialization.func1()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/civisibility_features.go:137 +0x234
>     sync.(*Once).doSlow()
>         /home/runner/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.24.2.linux-amd64.tar.gz/go/src/sync/once.go:78 +0xe1
>     sync.(*Once).Do()
>         /home/runner/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.24.2.linux-amd64.tar.gz/go/src/sync/once.go:69 +0x44
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations.ensureAdditionalFeaturesInitialization()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/civisibility_features.go:117 +0x78
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations.EnsureCiVisibilityInitialization.internalCiVisibilityInitialization.func2.1()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/civisibility.go:99 +0x38
>   
>   Goroutine 75 (finished) created at:
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations.EnsureCiVisibilityInitialization.internalCiVisibilityInitialization.func2()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/civisibility.go:99 +0x2e4
>     sync.(*Once).doSlow()
>         /home/runner/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.24.2.linux-amd64.tar.gz/go/src/sync/once.go:78 +0xe1
>     sync.(*Once).Do()
>         /home/runner/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.24.2.linux-amd64.tar.gz/go/src/sync/once.go:69 +0x44
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations.internalCiVisibilityInitialization()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/civisibility.go:60 +0x5b
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations.EnsureCiVisibilityInitialization()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/civisibility.go:44 +0x18
>     gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/integrations/gotesting.instrumentTestingM()
>         /home/runner/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.72.2/internal/civisibility/integrations/gotesting/instrumentation_orchestrion.go:44 +0xa4
>     testing.(*M).Run()
>         <generated>:2 +0x4a
>     main.main()
>         _testmain.go:45 +0x164
>   ==================
>   testing: race detected outside of test execution
>   ```
> </details> 

 _Originally posted by @pink in [#3413](https://github.com/DataDog/dd-trace-go/issues/3413#issuecomment-2857508929)_

--- a/internal/civisibility/integrations/civisibility_features.go
+++ b/internal/civisibility/integrations/civisibility_features.go
@@ -67,6 +67,7 @@ var (
 func ensureSettingsInitialization(serviceName string) {
 	settingsInitializationOnce.Do(func() {
 		log.Debug("civisibility: initializing settings")
+		defer log.Debug("civisibility: settings initialization complete")
 
 		// Create the CI Visibility client
 		ciVisibilityClient = net.NewClientWithServiceName(serviceName)
@@ -89,21 +90,24 @@ func ensureSettingsInitialization(serviceName string) {
 
 		// Get the CI Visibility settings payload for this test session
 		ciSettings, err := ciVisibilityClient.GetSettings()
-		if err != nil {
+		if err != nil || ciSettings == nil {
 			log.Error("civisibility: error getting CI visibility settings: %v", err)
-		} else if ciSettings != nil {
-			ciVisibilitySettings = *ciSettings
+			log.Debug("civisibility: no need to wait for the git upload to finish")
+			// Enqueue a close action to wait for the upload to finish before finishing the process
+			PushCiVisibilityCloseAction(func() {
+				<-uploadChannel
+			})
+			return
 		}
 
 		// check if we need to wait for the upload to finish and repeat the settings request or we can just continue
-		if ciVisibilitySettings.RequireGit {
+		if ciSettings.RequireGit {
 			log.Debug("civisibility: waiting for the git upload to finish and repeating the settings request")
 			<-uploadChannel
 			ciSettings, err = ciVisibilityClient.GetSettings()
 			if err != nil {
 				log.Error("civisibility: error getting CI visibility settings: %v", err)
-			} else if ciSettings != nil {
-				ciVisibilitySettings = *ciSettings
+				return
 			}
 		} else {
 			log.Debug("civisibility: no need to wait for the git upload to finish")
@@ -112,6 +116,35 @@ func ensureSettingsInitialization(serviceName string) {
 				<-uploadChannel
 			})
 		}
+
+		// check if we need to disable EFD because known tests is not enabled
+		if !ciSettings.KnownTestsEnabled {
+			// "known_tests_enabled" parameter works as a kill-switch for EFD, so if “known_tests_enabled” is false it
+			// will disable EFD even if “early_flake_detection.enabled” is set to true (which should not happen normally,
+			// the backend should disable both of them in that case)
+			ciSettings.EarlyFlakeDetection.Enabled = false
+		}
+
+		// check if flaky test retries is disabled by env-vars
+		if ciSettings.FlakyTestRetriesEnabled && !internal.BoolEnv(constants.CIVisibilityFlakyRetryEnabledEnvironmentVariable, true) {
+			log.Warn("civisibility: flaky test retries was disabled by the environment variable")
+			ciSettings.FlakyTestRetriesEnabled = false
+		}
+
+		// check if test management is disabled by env-vars
+		if ciSettings.TestManagement.Enabled && !internal.BoolEnv(constants.CIVisibilityTestManagementEnabledEnvironmentVariable, true) {
+			log.Warn("civisibility: test management was disabled by the environment variable")
+			ciSettings.TestManagement.Enabled = false
+		}
+
+		// overwrite the test management attempt to fix retries with the env var if set
+		testManagementAttemptToFixRetriesEnv := internal.IntEnv(constants.CIVisibilityTestManagementAttemptToFixRetriesEnvironmentVariable, -1)
+		if testManagementAttemptToFixRetriesEnv != -1 {
+			ciSettings.TestManagement.AttemptToFixRetries = testManagementAttemptToFixRetriesEnv
+		}
+
+		// set the ciVisibilitySettings with the settings from the backend
+		ciVisibilitySettings = *ciSettings
 	})
 }
 
@@ -119,7 +152,12 @@ func ensureSettingsInitialization(serviceName string) {
 func ensureAdditionalFeaturesInitialization(serviceName string) {
 	additionalFeaturesInitializationOnce.Do(func() {
 		log.Debug("civisibility: initializing additional features")
-		ensureSettingsInitialization(serviceName)
+		defer log.Debug("civisibility: additional features initialization complete")
+
+		// get a copy of the settings instance
+		currentSettings := *GetSettings()
+
+		// if there's no ciVisibilityClient then we don't need to do anything
 		if ciVisibilityClient == nil {
 			return
 		}
@@ -150,13 +188,26 @@ func ensureAdditionalFeaturesInitialization(serviceName string) {
 			additionalTags[key] = value
 		}
 
+		// if flaky test retries is enabled then let's load the flaky retries settings
+		if currentSettings.FlakyTestRetriesEnabled {
+			totalRetriesCount := (int64)(internal.IntEnv(constants.CIVisibilityTotalFlakyRetryCountEnvironmentVariable, DefaultFlakyTotalRetryCount))
+			retryCount := (int64)(internal.IntEnv(constants.CIVisibilityFlakyRetryCountEnvironmentVariable, DefaultFlakyRetryCount))
+			ciVisibilityFlakyRetriesSettings = FlakyRetriesSetting{
+				RetryCount:               retryCount,
+				TotalRetryCount:          totalRetriesCount,
+				RemainingTotalRetryCount: totalRetriesCount,
+			}
+			log.Debug("civisibility: automatic test retries enabled [retryCount: %v, totalRetryCount: %v]", retryCount, totalRetriesCount)
+		}
+
 		// wait group to wait for all the additional features to be loaded
 		var wg sync.WaitGroup
 
-		wg.Add(1)
-		go func() {
-			// if early flake detection is enabled then we run the known tests request
-			if ciVisibilitySettings.KnownTestsEnabled {
+		// if early flake detection is enabled then we run the known tests request
+		if currentSettings.KnownTestsEnabled {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
 				ciEfdData, err := ciVisibilityClient.GetKnownTests()
 				if err != nil {
 					log.Error("civisibility: error getting CI visibility known tests data: %v", err)
@@ -164,41 +215,14 @@ func ensureAdditionalFeaturesInitialization(serviceName string) {
 					ciVisibilityKnownTests = *ciEfdData
 					log.Debug("civisibility: known tests data loaded.")
 				}
-			} else {
-				// "known_tests_enabled" parameter works as a kill-switch for EFD, so if “known_tests_enabled” is false it
-				// will disable EFD even if “early_flake_detection.enabled” is set to true (which should not happen normally,
-				// the backend should disable both of them in that case)
-				ciVisibilitySettings.EarlyFlakeDetection.Enabled = false
-			}
-			wg.Done()
-		}()
+			}()
+		}
 
-		wg.Add(1)
-		go func() {
-			// if flaky test retries is enabled then let's load the flaky retries settings
-			if ciVisibilitySettings.FlakyTestRetriesEnabled {
-				flakyRetryEnabledByEnv := internal.BoolEnv(constants.CIVisibilityFlakyRetryEnabledEnvironmentVariable, true)
-				if flakyRetryEnabledByEnv {
-					totalRetriesCount := (int64)(internal.IntEnv(constants.CIVisibilityTotalFlakyRetryCountEnvironmentVariable, DefaultFlakyTotalRetryCount))
-					retryCount := (int64)(internal.IntEnv(constants.CIVisibilityFlakyRetryCountEnvironmentVariable, DefaultFlakyRetryCount))
-					ciVisibilityFlakyRetriesSettings = FlakyRetriesSetting{
-						RetryCount:               retryCount,
-						TotalRetryCount:          totalRetriesCount,
-						RemainingTotalRetryCount: totalRetriesCount,
-					}
-					log.Debug("civisibility: automatic test retries enabled [retryCount: %v, totalRetryCount: %v]", retryCount, totalRetriesCount)
-				} else {
-					log.Warn("civisibility: flaky test retries was disabled by the environment variable")
-					ciVisibilitySettings.FlakyTestRetriesEnabled = false
-				}
-			}
-			wg.Done()
-		}()
-
-		wg.Add(1)
-		go func() {
-			// if ITR is enabled then we do the skippable tests request
-			if ciVisibilitySettings.TestsSkipping {
+		// if ITR is enabled then we do the skippable tests request
+		if currentSettings.TestsSkipping {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
 				// get the skippable tests
 				correlationID, skippableTests, err := ciVisibilityClient.GetSkippableTests()
 				if err != nil {
@@ -208,35 +232,23 @@ func ensureAdditionalFeaturesInitialization(serviceName string) {
 					setAdditionalTags(constants.ItrCorrelationIDTag, correlationID)
 					ciVisibilitySkippables = skippableTests
 				}
-			}
-			wg.Done()
-		}()
+			}()
+		}
 
-		wg.Add(1)
-		go func() {
-			// if test management is enabled then we check if it was disabled by the environment variable
-			if ciVisibilitySettings.TestManagement.Enabled {
-				testManagementEnabledByEnv := internal.BoolEnv(constants.CIVisibilityTestManagementEnabledEnvironmentVariable, true)
-				testManagementAttemptToFixRetriesEnv := internal.IntEnv(constants.CIVisibilityTestManagementAttemptToFixRetriesEnvironmentVariable, -1)
-				if testManagementEnabledByEnv {
-					if testManagementAttemptToFixRetriesEnv != -1 {
-						ciVisibilitySettings.TestManagement.AttemptToFixRetries = testManagementAttemptToFixRetriesEnv
-					}
-
-					testManagementTests, err := ciVisibilityClient.GetTestManagementTests()
-					if err != nil {
-						log.Error("civisibility: error getting CI visibility test management tests: %v", err)
-					} else if testManagementTests != nil {
-						ciVisibilityTestManagementTests = *testManagementTests
-						log.Debug("civisibility: test management loaded [attemptToFixRetries: %v]", ciVisibilitySettings.TestManagement.AttemptToFixRetries)
-					}
-				} else {
-					ciVisibilitySettings.TestManagement.Enabled = false
-					log.Warn("civisibility: test management was disabled by the environment variable")
+		// if test management is enabled then we do the test management request
+		if currentSettings.TestManagement.Enabled {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				testManagementTests, err := ciVisibilityClient.GetTestManagementTests()
+				if err != nil {
+					log.Error("civisibility: error getting CI visibility test management tests: %v", err)
+				} else if testManagementTests != nil {
+					ciVisibilityTestManagementTests = *testManagementTests
+					log.Debug("civisibility: test management loaded [attemptToFixRetries: %v]", currentSettings.TestManagement.AttemptToFixRetries)
 				}
-			}
-			wg.Done()
-		}()
+			}()
+		}
 
 		// wait for all the additional features to be loaded
 		wg.Wait()

